pkg/metrics/register.go

```diff
@@ -0,0 +1,76 @@
+package metrics
+
+import (
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/zondax/golem/pkg/metrics/collectors"
+)
+
+type collectorRegister func(name, help string, labels []string, handler MetricHandler) (prometheus.Collector, error)
+
+var (
+	collectorRegisterMap = map[string]collectorRegister{
+		collectors.CounterType:   registerCounter,
+		collectors.HistogramType: registerHistogram,
+		collectors.GaugeType:     registerGauge,
+	}
+)
+
+func (t *taskMetrics) RegisterMetric(name string, help string, labels []string, handler MetricHandler) error {
+	var metric prometheus.Collector
+
+	registerFn, ok := collectorRegisterMap[handler.Type()]
+	if !ok {
+		return fmt.Errorf("unsupported metric type")
+	}
+
+	metric, err := registerFn(name, help, labels, handler)
+	if err != nil {
+		return err
+	}
+
+	if err = prometheus.Register(metric); err != nil {
+		return err
+	}
+
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	t.metrics[name] = MetricDetail{Collector: metric, Handler: handler}
+
+	return nil
+}
+
+func registerCounter(name, help string, labels []string, handler MetricHandler) (prometheus.Collector, error) {
```

Comment:
Should we check here that the MetricHandler is Counter type as we do with registerHistogram method?

Reply:
Yes, we check this here `if _, ok := handler.(*collectors.Counter); !ok {`